### PR TITLE
[internal/scrapertest] Add CompareOption for ignoring attribute values

### PR DESCRIPTION
**Pull request by djaglowski**
Attribute values are sometimes derived from dynamic environments.
In order to support integration tests, this change provides a new
implementation of scrapertest.CompareOption that can be configured
to ignore specific attribute values.

The CompareOption requires that the attribute name is specified.
Optionally, one or more metric names may be specified as well. If
any metric names are specified, than the attribute value will only
be ignored on those metrics.

### CODE REVIEW

**Comment by djaglowski** [comment]
I've converted this back to a draft in order to address a couple issues I realized.

1. The test cases could be more thorough, but they're difficult to reasonable about in the current format. I'm going to refactor tests on a separate PR, and then rebase onto that. 
2. There's an important case that is not yet handled. Basically, because we are wiping out attributes, it's currently possible for data points to lose their uniqueness. For example, if we have two points with a metric, with attributes `{ "same": "same", "diff": "one" }` and `{ "same": "same", "diff": "two" }`, and then we ignore the value of `"diff"`, then these points are no longer differentiated from each other, and comparison will often fail in this case. A solution to this is tricky, but I think the best thing would be to sort all data points based on non-attribute values (start/end timestamp, value), and then assign attribute values from a predictable sequence. If any data points in the metric are identical except for attribute values, then they will at least be in a predictable sequence, and tests can pass.

**Comment by djaglowski** [comment]
I've rebased this onto #6621 and added some additional clarification to the test suite.

File: internal/scrapertest/mask.go [hunk 1]
```diff
@@ -61,3 +47,78 @@ func maskDataPointSliceValues(dataPoints pdata.NumberDataPointSlice) {
...
+		attribute, ok := attributes.Get(attributeName)
+		if ok {
+			switch attribute.Type() {
+			case pdata.AttributeValueTypeString: // ok
+			default:
+				panic(fmt.Sprintf("data type not supported: %s", attribute.Type()))
+			}
+		}
+
+		attributes.UpsertString(attributeName, "")
```
Comment (dmitryax) [inline internal/scrapertest/mask.go:122]: Do we need `Upsert`? Isn't it expected that the attribute is always set? Metrics builder currently doesn't support recording datapoints without providing all attributes specified in metadata.yaml.

File: internal/scrapertest/compare.go [hunk 2]
```diff
@@ -105,14 +106,24 @@ func CompareNumberDataPointSlices(expected, actual pdata.NumberDataPointSlice) e
 		return fmt.Errorf("length of datapoints don't match")
 	}
 
+	numPoints := expected.Len()
+
+	// Keep track of which points have already been used as a match
+	// so that each point can only be matched once
+	once := make(map[int]bool, numPoints)
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply internal/scrapertest/compare.go:113]: There are a couple minor correctness and readability improvements here. I've split them off into #7305 and will rebase this PR onto that branch.

File: internal/scrapertest/mask.go [hunk 3]
```diff
@@ -61,3 +47,78 @@ func maskDataPointSliceValues(dataPoints pdata.NumberDataPointSlice) {
...
+func maskMetricSliceAttributeValues(metrics pdata.MetricSlice, attributeName string, metricNames ...string) {
+	if len(metricNames) == 0 {
+		for i := 0; i < metrics.Len(); i++ {
+			dps := getDataPointSlice(metrics.At(i))
+			maskDataPointSliceAttributeValues(dps, attributeName)
+
+			// If attribute values are ignored, some data points may become
+			// indistinguishable from each other, but sorting by value allows
+			// for a reasonably thorough comparison and a deterministic outcome.
+			dps.Sort(func(a, b pdata.NumberDataPoint) bool {
```
[the remark this answers is not in the document]
Comment (djaglowski) [reply internal/scrapertest/mask.go:82]: Good catch. Good suggestion.

File: internal/scrapertest/mask.go (hunk 1 above)
Comment (djaglowski) [reply internal/scrapertest/mask.go:122]: I was not aware, but it is a good point even if used without metadata.yaml. We should only ignore if the value is present. If entirely absent, this is notable difference in expectation and comparison should fail.

**Comment by djaglowski** [comment]
@dmitryax I believe this one is all set now.

**Review by dmitryax — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

**Pull request merged**